libfwupdplugin/fu-smbios.c

```diff
@@ -89,6 +93,76 @@ fu_smbios_convert_dt_string (FuSmbios *self, guint8 type, guint8 offset,
 	fu_smbios_convert_dt_value (self, type, offset, item->strings->len);
 }
 
+#if defined (__FreeBSD__) || defined (__DragonFly__)
+static void
+fu_smbios_kenv_sysctl_string (FuSmbios *self, guint8 type, guint8 offset,
+				const gchar *buf, gsize bufsz)
+{
+	FuSmbiosItem *item = g_ptr_array_index (self->items, type);
+
+	/* add to strtab */
+	g_ptr_array_add (item->strings, g_strndup (buf, bufsz));
+	fu_smbios_convert_dt_value (self, type, offset, item->strings->len);
+}
+
+static gboolean
+fu_smbios_kenv_lookup (const char *sminfo, gchar *buf, gsize bufsz)
+{
+	if (kenv (KENV_GET, sminfo, buf, bufsz - 1) == -1) {
+		g_debug ("cannot get DMI request for %s", sminfo);
+		return FALSE;
+	}
+	return TRUE;
+}
+
+static void
+fu_smbios_convert_kenv_string (FuSmbios *self, guint8 type, guint8 offset,
+				const char *sminfo)
+{
+	/* Maximum value length - 128 */
+	gchar buf[128] = { '\0' };
+	gsize bufsz = sizeof(buf);
+
+	if (fu_smbios_kenv_lookup (sminfo, buf, bufsz))
+		fu_smbios_kenv_sysctl_string (self, type, offset, buf, bufsz);
+}
+
+static gboolean
+fu_smbios_setup_from_kenv (FuSmbios *self, GError **error)
+{
+	/* add all four faked structures */
+	for (guint i = 0; i < FU_SMBIOS_STRUCTURE_TYPE_LAST; i++) {
+		FuSmbiosItem *item = g_new0 (FuSmbiosItem, 1);
+		item->type = i;
+		item->buf = g_byte_array_new ();
+		item->strings = g_ptr_array_new_with_free_func (g_free);
+		g_ptr_array_add (self->items, item);
+	}
+
+	/* DMI:Manufacturer */
+	fu_smbios_convert_kenv_string (self, FU_SMBIOS_STRUCTURE_TYPE_SYSTEM, 0x04,
+				     "smbios.bios.vendor");
+
+	/* DMI:BiosVersion */
+	fu_smbios_convert_kenv_string (self, FU_SMBIOS_STRUCTURE_TYPE_BIOS, 0x05,
+				     "smbios.bios.version");
+
+	/* DMI:Family */
+	fu_smbios_convert_kenv_string (self, FU_SMBIOS_STRUCTURE_TYPE_SYSTEM, 0x1a,
+				     "smbios.system.family");
+
+	/* DMI:ProductName */
+	fu_smbios_convert_kenv_string (self, FU_SMBIOS_STRUCTURE_TYPE_SYSTEM, 0x05,
+				     "smbios.planar.product");
+
+	/* DMI:BaseboardManufacturer */
+	fu_smbios_convert_kenv_string (self, FU_SMBIOS_STRUCTURE_TYPE_BASEBOARD,
+				     0x04, "smbios.planar.maker");
+
+	return TRUE;
```

Comment:
if none of these strings setup properly this will just be silently hidden right now, but I think it should be a failure if none get setup.  What do you think about only returning TRUE if at least one sets up?

Reply:
Done.